Changed adding chems to plants into a simple lerp

genderGeometries opened the pull request:
<!-- Please read these guidelines before opening your PR: https://docs.spacestation14.io/en/getting-started/pr-guideline -->
<!-- The text between the arrows are comments - they will not be visible on your PR. -->

## About the PR
<!-- What did you change in this PR? -->
Removed `PotencyDivisor` from the calculation for adding chems to crops on harvest, replacing it with a lerp. 
## Why / Balance
<!-- Why was it changed? Link any discussions or issues here. Please discuss how this would affect game balance. -->
The old calculation is unintuitive and makes it difficult to add new crops. Looking through `seeds.yml`, there are a wide number of crops with `Max` chem values that will never be reached. These values have been changed to match attainable values in `seeds.yml`, so there should be no change in game play.
(A future PR will attempt a rebalance of nutriment, vitamin, and cooking chems now that the numbers are more straightforward.)
## Technical details
<!-- If this is a code change, summarize at high level how your new code works. This makes it easier to review. -->
Chems in crops are currently calculated in `BotanySystem.Produce.cs` as:
`Clamp(Min + Potency/PotencyDivisor, Min, Max)`
This has been changed to:
`Clamp(Lerp(Min, Max, Potency/50), Min, Max)`
50 is used instead of 100 because 50 is the current max potency (`PotencyLimit`) in `RobustHarvest.cs`. I considered changing `PotencyLimit` to 100 and `PotencyIncrease` to 6, and may do so in a future PR.
PotencyDivisor has been deleted from all areas it is mentioned, and min and max values of crops has been changed to values attainable in-game. Contact me if you'd like access to the spreadsheet. :)

## Media
<!-- 
PRs which make ingame changes (adding clothing, items, new features, etc) are required to have media attached that showcase the changes.
Small fixes/refactors are exempt.
Any media may be used in SS14 progress reports, with clear credit given.

If you're unsure whether your PR will require media, ask a maintainer.

Check the box below to confirm that you have in fact seen this (put an X in the brackets, like [X]):
-->

![Untitled](https://github.com/space-wizards/space-station-14/assets/159584039/12cc5a6a-28a3-42c2-ad66-63dd560cd07f)
![2](https://github.com/space-wizards/space-station-14/assets/159584039/5ba95b96-65b2-4d62-b4e4-dac57c201de5)
![3](https://github.com/space-wizards/space-station-14/assets/159584039/33a913bb-9a82-464d-8fc7-2a3599444996)
![botanysheet](https://github.com/space-wizards/space-station-14/assets/159584039/d919583d-5124-40dd-9daf-00c3f0fbb60b)

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes
<!--
List any breaking changes, including namespace, public class/method/field changes, prototype renames; and provide instructions for fixing them. This will be pasted in #codebase-changes.
-->
The data field `PotencyDivisor` has been removed from `SeedPrototype.cs`.
**Changelog**
<!--
Make players aware of new features and changes that could affect how they play the game by adding a Changelog entry. Please read the Changelog guidelines located at: https://docs.spacestation14.io/en/getting-started/pr-guideline#changelog
-->
:cl:
- tweak: Crop chemicals are calculated in a more intuitive way.

UbaserB commented:
why do lemons contain milk i thought it was only lemoons

genderGeometries commented:
> why do lemons contain milk i thought it was only lemoons

Oh haha, I put milk in the lemon line on my spreadsheet purely out of laziness and because I'm planning a re-balancing PR to follow up #25366. Milk has not been added to lemons.

botanySupremist commented:
Does this mean the maximum potency that can change a plant's reagents is 50?

genderGeometries commented:
> Does this mean the maximum potency that can change a plant's reagents is 50?

In `RobustHarvest.cs`, not changed in this PR, if potency is at 50 it does not increase further. I am not aware of a way for potency to go over 50 currently; if there is a way potency can be increased beyond 50, I will have to reconsider this PR.

ViceEmargo commented:
> > Does this mean the maximum potency that can change a plant's reagents is 50?
> 
> In `RobustHarvest.cs`, not changed in this PR, if potency is at 50 it does not increase further. I am not aware of a way for potency to go over 50 currently; if there is a way potency can be increased beyond 50, I will have to reconsider this PR.

Potency has go over 50 by mutation.
Although the highest I've been able to get plants through dumptruck loads of mutagen is 58.
I believe when mutagen is combined with Left4Zed, it can be higher, but I honestly doubt anyone **EVER** has been able to hit anywhere near 100 in a non-test environment. A plant would pick up so much other baggage you wouldn't want long before you could obtain near 100 potency.

Aisu9 commented:
50 is the maximum a plant can reach using Robust Harvest, the plant can reach much higher with mutations, you gotta be luckiest nerds to even have one plant reaching 100 potency by the end of the shift without any debilitating mutations

melanoTurbo commented:
if you 
- want to spend 30-60 minutes at botany 
- have a bucket or two of mutagen
- looked into mutation code and know what you are doing
- playing on an older build that doesnt have clipping nerf
then you can reach 100 potency

melanoTurbo commented:
> > > Does this mean the maximum potency that can change a plant's reagents is 50?
> > 
> > 
> > In `RobustHarvest.cs`, not changed in this PR, if potency is at 50 it does not increase further. I am not aware of a way for potency to go over 50 currently; if there is a way potency can be increased beyond 50, I will have to reconsider this PR.
> 
> Potency has the potential to go over 50 by mutation. Although the highest I've been able to get plants through dumptruck loads of mutagen is 58. I believe when mutagen is combined with Left4Zed, it can be higher, but I honestly doubt anyone **EVER** has been able to hit anywhere near 100 in a non-test environment. A plant would pick up so much other baggage you wouldn't want long before you could obtain near 100 potency.

you dont need to dump a lot of mutagen into one plant to reach 100 potency. you just add 1-5 units, clip it after a while, see potency on the seeds bag and discard it if potency hasnt increased

and you repeat it for the rest of the round to get 100 potency a minute befor evac. that has been done a lot of times before clipping nerf, but it is extremely boring and doesnt have any gameplay value worth the time

Kadeo64 commented:
yeah any potency above 50 is basically a fever dream. unless you're a hyper robust botany shitter micromanaging the system and complaining about lag spikes because it throws off your perfect gameplay you will realistically never get above 50 potency outside of plants you uprooted anyway because they didn't mutate into a new plant

genderGeometries commented:
Thanks everyone for the feedback. I removed the clamp in the chem calculation, so mutations should have beeg chems.

Additionally, after testing the mutation system, I found that mutating potency always sets it to 44, 20% of the way between the 30 and 100 values it is supposed to pick from. I confirmed the 20% by changing the potency mutation range to 40-100, and it went to 52, also 20% of the range. This bug is outside of the scope of this PR, but I will turn my eye to it next.

Partmedia commented on Content.Server/Botany/Systems/BotanySystem.Produce.cs:
```diff
@@ -15,10 +15,13 @@ public void ProduceGrown(EntityUid uid, ProduceComponent produce)
+            if (seed.Potency > 0)
+            {
+                amount += FixedPoint2.New(MathHelper.Lerp(quantity.Min, quantity.Max, seed.Potency / 50));
+                // Adds an amount, depending on the potency. The maximum is 50 as per RobustHarvest.cs,
+                // but mutations (MutationSystem.cs) can cause it to go over the max, up to 100.
```
Comments should be above the lines that they are commmenting.

Partmedia commented on Content.Server/Botany/Systems/BotanySystem.Produce.cs:
```diff
@@ -15,10 +15,13 @@ public void ProduceGrown(EntityUid uid, ProduceComponent produce)
-                amount += FixedPoint2.New(seed.Potency / quantity.PotencyDivisor);
-            amount = FixedPoint2.New(MathHelper.Clamp(amount.Float(), quantity.Min, quantity.Max));
+            var amount = FixedPoint2.Zero;
+            if (seed.Potency > 0)
```
Now the minimum reagent value won't be respected. Notice how in the old code, the minimum is is always added to the product, even if the potency is zero. Is this really intended?

genderGeometries replied:
I originally had amount start at Min and Lerp from 0 to Max. I did not know potency could be 0; I will fix this, thank you.

Partmedia commented on Content.Server/Botany/Systems/BotanySystem.Produce.cs:
```diff
@@ -15,10 +15,13 @@ public void ProduceGrown(EntityUid uid, ProduceComponent produce)
+            var amount = FixedPoint2.Zero;
+            if (seed.Potency > 0)
+            {
+                amount += FixedPoint2.New(MathHelper.Lerp(quantity.Min, quantity.Max, seed.Potency / 50));
```
Why `+=` where a `=` would suffice?

Partmedia commented on Content.Server/Botany/Systems/BotanySystem.Produce.cs:
```diff
@@ -15,10 +15,13 @@ public void ProduceGrown(EntityUid uid, ProduceComponent produce)
+            if (seed.Potency > 0)
+            {
+                amount += FixedPoint2.New(MathHelper.Lerp(quantity.Min, quantity.Max, seed.Potency / 50));
+                // Adds an amount, depending on the potency. The maximum is 50 as per RobustHarvest.cs,
```
Now the maximum isn't correctly honored. This should clamp at maximum potency. This new denominator should be greater than 50, because otherwise there would be no increase in chemicals despite increases in potency.

genderGeometries replied:
I have the denominator at 50 because it is the maximum when using Robust Harvest in-game, and I have seen botanists (such as the comments above) talk about 50 as the "practical" maximum. It is possible to get potency above 50, and in my tests the `Lerp` function successfully "extrapolated". (By which I mean, when potency is above 50, there is an increase in chemicals above the "maximum" value.) I think being able to use mutation to get potency/chemicals "above the maximum" is a fun sort of chaos in the game, and does not need to be removed in this PR. Would it be a reasonable compromise to have it clamp at 100 potency (2x"max")?

Partmedia replied:
So then `max` is not truly a maximum, and the datafield name should reflect that. Perhaps `robustHarvestMax`, and then un-hard-code the robust harvest limit.

genderGeometries replied:
These are good ideas, thank you!

genderGeometries replied:
Did you have a strategy in mind for un-hard-coding the robust harvest limit? I'm unsure how to reference PotencyLimit from the BotanySystem class.

Partmedia reviewed (changes requested):
Thank you for your PR. I do feel that this would help with making plant number balancing less confusing.

> tweak: Crop chemicals are calculated in a more intuitive way.

Please focus on the user-visible changes. If there are no user-visible changes, this should not be in the Changelog per the [Changelog guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).

In your breaking changes, you need to explain how forks should fix this. You will need to discuss what needs to change in YAML, and how to adjust the new `Max` field. Put yourself in the shoes of someone who is updating and needs to know how to convert their custom plants. Perhaps including the formula you used to rebalance the values without any in-game change would help.

Please also see the in-line comments.

genderGeometries commented:
In SeedPrototype.cs:
- Marked out any places where Max is mentioned in code and replaced it with PotencyAddend
- Changed Max in SeedPrototype.cs to PotencyAddend
- Added summary to PotencyAddend
In BotanySystem.Produce.cs
- `amount` starts set as Min
- Lerp from 0 to PotencyAddend is added to `amount`
- Moved the comment
In seeds.yml:
- Changed max: to potencyaddend:
- subtracted min from max to get potencyaddend value

Still working on how to reference PotencyLimit from BotanySystem.Produce.cs in order to un-hard-code the 50...

mehx3 commented:
> Thanks everyone for the feedback. I removed the clamp in the chem calculation, so mutations should have beeg chems.
> 
> Additionally, after testing the mutation system, I found that mutating potency always sets it to 44, 20% of the way between the 30 and 100 values it is supposed to pick from. I confirmed the 20% by changing the potency mutation range to 40-100, and it went to 52, also 20% of the range. This bug is outside of the scope of this PR, but I will turn my eye to it next.

Potency modifying mutation changes potency in range from 30 to 100, in increments of 12%: 30, 44, 58, 72, 86, 100. The higher it gets, the lower the chance for it's increase is.

slarticodefast commented:
Are you coming back to this?

genderGeometries commented:
> Are you coming back to this?

Ah! I will make plans to look at cleaning this up for another go in a week when I do not have COVID.

slarticodefast commented:
Alright, no need to hurry, just wanted to check if this PR is still active. I wish you a good recovery.
